hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/IdClassEmbeddable.java

```diff
@@ -0,0 +1,780 @@
+			attribute.breakDownJdbcValues( attributeValue, valueConsumer, session );
+		} );
+	}
+
+	@Override
+	public SelectableMapping getSelectable(int columnIndex) {
+		return selectableMappings.getSelectable( columnIndex );
+	}
+
+	@Override
+	public int forEachSelectable(SelectableConsumer consumer) {
+		return selectableMappings.forEachSelectable( 0, consumer );
+	}
+
+	@Override
+	public int forEachSelectable(int offset, SelectableConsumer consumer) {
+		return selectableMappings.forEachSelectable( offset, consumer );
+	}
+
+	@Override
+	public int getJdbcTypeCount() {
+		return selectableMappings.getJdbcTypeCount();
+	}
+
+	@Override
+	public List<JdbcMapping> getJdbcMappings() {
+		return selectableMappings.getJdbcMappings();
+	}
+
+	@Override
+	public int forEachJdbcValue(
+			Object value,
+			Clause clause,
+			int offset,
+			JdbcValuesConsumer valuesConsumer,
+			SharedSessionContractImplementor session) {
+		int span = 0;
+
+		for ( int i = 0; i < attributeMappings.size(); i++ ) {
+			final AttributeMapping attributeMapping = attributeMappings.get( i );
+			if ( attributeMapping instanceof PluralAttributeMapping ) {
+				continue;
+			}
+			final Object o = attributeMapping.getPropertyAccess().getGetter().get( value );
+			span += attributeMapping.forEachJdbcValue( o, clause, span + offset, valuesConsumer, session );
+		}
+		return span;
+	}
+
+	@Override
+	public Object disassemble(Object value, SharedSessionContractImplementor session) {
+		final List<AttributeMapping> attributeMappings = getAttributeMappings();
+
+		// todo (6.0) : reduce to-one values to id here?
+		final Object[] result = new Object[ attributeMappings.size() ];
+		for ( int i = 0; i < attributeMappings.size(); i++ ) {
+			final AttributeMapping attributeMapping = attributeMappings.get( i );
+			Object o = attributeMapping.getPropertyAccess().getGetter().get( value );
+			result[i] = attributeMapping.disassemble( o, session );
+		}
+
+		return result;
+	}
+
+	@Override
+	public int forEachJdbcType(int offset, IndexedConsumer<JdbcMapping> action) {
+		return selectableMappings.forEachSelectable(
+				offset,
+				(index, selectable) -> action.accept( index, selectable.getJdbcMapping() )
+		);
+	}
+
+	@Override
+	public <T> DomainResult<T> createDomainResult(NavigablePath navigablePath, TableGroup tableGroup, String resultVariable, DomainResultCreationState creationState) {
+		throw new NotYetImplementedFor6Exception( getClass() );
+	}
+
+	@Override
+	public EmbeddableMappingType createInverseMappingType(
+			EmbeddedAttributeMapping valueMapping,
+			TableGroupProducer declaringTableGroupProducer,
+			SelectableMappings selectableMappings,
+			MappingModelCreationProcess creationProcess) {
+		return new EmbeddableMappingTypeImpl(
+				valueMapping,
+				declaringTableGroupProducer,
+				selectableMappings,
+				this,
+				creationProcess
+		);
+	}
+
+
+
+
+
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// init
```

Comment:
I think we should try to move most of this logic into some abstract base class if possible, as that is AFAICS duplicated from EmbeddableMappingTypeImpl. It appears that the same/similar duplicate code is also used in VirtualIdEmbeddable

Reply:
Yep, as mentioned in my commit comments ;)